proposals/0008-package-manager-testing.md

```diff
@@ -0,0 +1,194 @@
+
+### Command Output
+
+Executing a test from the terminal will produce user-readable output.
+This should incorporate colorization and other formatting
+similar to other testing tools
+to indicate the success and failure of different tests.
+For example:
+
+    $ swift test --output module
+    Running tests for PackageX (x/100)
+    .........x.....x...................
+
+    Completed
+    Elapsed time: 0.2s
+
+    98 Success
+     2 Failure
+     1 Warning
+
+    FAILURE: Tests/TestsA.swift:24 testFoo()
+    XCTAssertTrue expected true, got false
+
+    FAILURE: Tests/TestsB.swift:10 testBar()
+    XCTAssertEqual
+
+    WARNING: Tests/TestsC.swift:1
+    "Some Warning"
+
+An additional option may be passed to the testing command
+to output JUnit-style XML or other formats that can be integrated
+with continuous integration (CI) and other systems.
+
+### Backwards Compatibility
+
+In order to accomodate existing packages,
+we will allow test module targets and their targets
+to be overridden in the `Package.swift` manifest file.
+However, this functionality will likely not be implemented
+in the initial release of this feature,
+and instead be added at a later point in time.
+
+### Automatic Dependency Determination
+
+Testing is important and it is important to make the barrier to testing
+as minimal as possible. Thus, by analyzing the names of test targets,
+we will automatically determine the most likely dependency of that test
+and accomodate accordingly.
+For example,
+a test for "Foo" will depend on compilation of the library target `Foo`.
+Any additional dependencies or dependencies that could not be automatically determined
+would need to be specified in a package manifest separately.
+
+### Debug / Release Configuration
+
+Although tests built in debug configuration
+are generally run against modules also build in debug configuration,
+it is sometimes necessary to specify the build configuration for tests separately.
+It is also sometimes necessary to explicitly specify this information for every build,
+such as when building in a release configuration to execute performance tests.
+We would like to eventually support these use cases,
+however this will not be present in the initial implementation of this feature.
+
+### Testability
+
+Swift can build modules with "testability",
+which allows tests to access entities with `internal` access control.
+Because it would be tedious for users to specify this requirement for tests,
+we intend to build debug builds with testability by default.
+
+### Test Frameworks
+
+Initially,
+the Swift Package Manager will use `XCTest` as its underlying test framework.
+
+However, testing is an evolving artform,
+so we'd like to support other approaches
+that might allow frameworks other than XCTest
+to be supported by the package manager.
+We expect that such an implementation would take the form of
+a Swift protocol that the package manager defines,
+which other testing frameworks can adopt.
```

Comment:
> @kylef’s proposal that the presence of a main.swift causes standalone test-binaries has the serious con IMO that any additional infrastructure people may want to use on top of swiftpm testing will be decoupled. If the minimum layer is a protocol, then we can build whatever test-runners the community wants on top of that.

Can you elaborate on what the problem is with things being decoupled? In my opinion, being decoupled would be a huge improvement.

I think it's fine that there is the XCTest protocol, but I really would like to have a way to not use it, even if that option is not default. Having the flexibility to write tests without any limitations or constraints imposed by the XCTest protocol would be beneficial. There are many different preferences on how testing can be done and I would hate to see a certain design being tightly coupled into the language and ecosystem. Tight coupling to a particular testing design might hamper Swift adoption in as-yet unforeseen applications.

As a proof of concept. I've build a simple runner (and unfortunately also a simple build tool for tests) which allows Spectre to work with SPM (although not integrated). This allows a `Tests/` directory where I can have a `main.swift` which runs the tests. https://github.com/kylef/spectre-build